### PR TITLE
Additions to the Lua API and scriptability interface

### PR DESCRIPTION
**Pull request by nashiora**
A new Scriptable class is added, which is now an optional parameter to Application::LoadScript. When the application sees one (when loading a script) it calls an initialization function before loading the script file. This allows a Scriptable to initialize global values and functions etc. before the script is actually run.

The application now also exposes a global include function, which operates relative to the skin's folder (NOT relative to the script calling the function.) This function will load in a script (without caching, since USC will reload scripts whenever the hell it wants so) by appending the string parameter to the skins directory and adding the .lua extension.

The title screen now allows the skin script to call the SongSelect, Settings and Exit functions rather than querying for them. This might not be the best solution, but any script that would ever want to be used would implement this just fine so it shouldn't be an issue. I like the added programmability of the title screen so I think it's worth doing this way. We can always revert to a polling system so long as it's NOT just in the mouse_pressed function, but that might be a discussion for another day.

The default skin now also implements basic .ini file parsing so that skins can more easily support their own configuration. I'd like to see more ability for custom input in skin scripts, more than just the controller for example and maybe an idea to request more direct access to inputs, but again that would mean changing the way a skin programs the screens and will definitely need to be done with more care.

### CODE REVIEW

File: Main/Application.cpp [hunk 1]
```diff
@@ -1113,6 +1127,23 @@ static int lSetGaugeColor(lua_State* L /*int colorIndex, int r, int g, int b*/)
 	return 0;
 }
 
+int lInclude(lua_State* L)
```
Comment (jerwuqu) [inline Main/Application.cpp]: Doesn't lua already have `require`?
https://www.lua.org/pil/8.1.html

Comment (nashiora) [reply Main/Application.cpp]: At the time I wasn't aware of `LUA_PATH` as anything more than an environment variable. I'll look at using that to configure `require` and see where it goes.

File: docs/source/globals.rst [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+Included files may return values; if so, these values are returned from include.
+The path parameter is expected to the a normal path relative to the skin folder, without
+the .lua file extension.
+
+It's good practice in many cases to return a table containing the values and functions
+you would like to export to someone who includes the script.
+See the ini.lua file in the default script.
+
+Included files are re-loaded every time they are include.
+USC is very library with reloading scripts, so include avoids caching anything.
```
[the remark this answers is not in the document]
Comment (nashiora) [reply docs/source/globals.rst]: fixed

File: Main/TitleScreen.cpp [hunk 3]
```diff
@@ -14,84 +14,88 @@
...
+		m_RegisterMemberFunction("SongSelect", m_Bind(&TitleScreen_Impl::lSongSelect));
+		m_RegisterMemberFunction("Settings", m_Bind(&TitleScreen_Impl::lSettings));
+		m_RegisterMemberFunction("Exit", m_Bind(&TitleScreen_Impl::lExit));
+
+		m_CreateGlobalObject("menu");
+	};
+
 	~TitleScreen_Impl()
 	{
+		// TODO(local): needs to handled by Scriptable, but I can't get it to have
```
[the remark this answers is not in the document]
Comment (nashiora) [reply Main/TitleScreen.cpp]: The initial reason for not putting in in Scriptable is I couldn't get the compiler or linker to be happy about `#include`ing the stuff I needed. Hopefully moving stuff to Scriptable will help, but I also would have to update everything in the game to work with it; I'll see what I can manage to make everything happy.

File: Main/TitleScreen.cpp [hunk 4]
```diff
@@ -14,84 +14,88 @@
...
 		return true;
 	}
+
+	virtual void m_InitScriptState()
+	{
+		m_RegisterMemberFunction("SongSelect", m_Bind(&TitleScreen_Impl::lSongSelect));
+		m_RegisterMemberFunction("Settings", m_Bind(&TitleScreen_Impl::lSettings));
+		m_RegisterMemberFunction("Exit", m_Bind(&TitleScreen_Impl::lExit));
+
+		m_CreateGlobalObject("menu");
```
[the remark this answers is not in the document]
Comment (nashiora) [reply Main/TitleScreen.cpp]: The whole API currently is meant to work around exactly that: a global object with member functions. It's definitely not self-documenting, and in the process of upgrading the class I'll have that be more clear, better documented, and an option vs global functions and values where needed.

File: docs/source/globals.rst [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+Loads a Lua file into the current environment.
+Included files may return values; if so, these values are returned from include.
+The path parameter is expected to the a normal path relative to the skin folder, without
+the .lua file extension.
+
+It's good practice in many cases to return a table containing the values and functions
+you would like to export to someone who includes the script.
+See the ini.lua file in the default script.
+
+Included files are re-loaded every time they are include.
```
[the remark this answers is not in the document]
Comment (nashiora) [reply docs/source/globals.rst]: Having to worry about clearing a cached script isn't something I think should be necessary, but if caching isn't an issue then it wouldn't be too bad.

File: Main/Application.cpp [hunk 6]
```diff
@@ -725,13 +725,18 @@ int Application::LoadImageJob(const String & path, Vector2i size, int placeholde
 	return ret;
 }
 
-lua_State* Application::LoadScript(const String & name)
+lua_State* Application::LoadScript(const String & name, Scriptable* scriptable)
```
[the remark this answers is not in the document]
Comment (nashiora) [reply Main/Application.cpp]: Yeah, I hated adding this dirty little bit but it kept me from having to move absolutely everything to a new system when most of it wasn't really necessary. I'll look around thru the other scriptable areas and make Scriptable a better common fit.

File: docs/source/globals.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (jerwuqu) [reply docs/source/globals.rst]: The requires are cached in the luaState so if you were to reload the skin itself, the skins' require caches wouldn't live through that. I suppose you'd need it if you had a skin that did some reloading of its own requires for a Lua-based config or something? Not really sure. In either case, the `require` function can be overridden to always clear cache as the default behavior whenever requiring something.

File: Main/TitleScreen.cpp [hunk 7]
```diff
@@ -14,84 +14,88 @@
...
 	bool Init()
 	{
 		m_guiStyle = g_commonGUIStyle;
-		CheckedLoad(m_lua = g_application->LoadScript("titlescreen"));
+		CheckedLoad(g_application->LoadScript("titlescreen", this));
 		g_gameWindow->OnMousePressed.Add(this, &TitleScreen_Impl::MousePressed);
 		return true;
 	}
+
+	virtual void m_InitScriptState()
```
[the remark this answers is not in the document]
Comment (nashiora) [reply Main/TitleScreen.cpp]: Should've been protected, missed that.

**Pull request closed without merging**